File: INTRODUCCION-POO/EJERCICIO-7/JAVA/Celular.java
import java.util.ArrayList;
import java.util.List;

class Celular {
    private List<Aplicacion> aplicaciones;
    private int espacioUsado;
    private int espacioTotal;
    private double bateria;

    public Celular() {
        this.aplicaciones = new ArrayList<>();
        this.espacioUsado = 0;
        this.espacioTotal = 1024;
        this.bateria = 100;
    }

    public void instalarAplicacion(String nombre, int tamaño) {
        if (aplicaciones.size() >= 20) {
            System.out.println("No se pueden instalar más de 20 aplicaciones.");
            return;
        }
        if (espacioUsado + tamaño > espacioTotal) {
            System.out.println("No hay suficiente espacio para instalar la aplicación.");
            return;
        }

        aplicaciones.add(new Aplicacion(nombre, tamaño));
        espacioUsado += tamaño;
        System.out.println("Aplicación " + nombre + " instalada con éxito.");
    }

    public void usarAplicacion(String nombre, int minutos) {
        if (bateria == 0) {
            System.out.println("Celular apagado. Por favor, cargue la batería.");
            return;
        }

        Aplicacion app = null;
        for (Aplicacion a : aplicaciones) {
            if (a.nombre.equals(nombre)) {
                app = a;
                break;
            }
        }

        if (app == null) {
            System.out.println("La aplicación no está instalada.");
            return;
        }

        double consumoBateria = 0;
        if (app.tamaño > 250) {
            consumoBateria = 0.05 * (minutos / 10);
        } else if (app.tamaño > 100) {
            consumoBateria = 0.02 * (minutos / 10);
        } else {
            consumoBateria = 0.01 * (minutos / 10);
        }

        bateria -= consumoBateria * 100;
        if (bateria < 0) {
            bateria = 0;
        }

        if (bateria == 0) {
            System.out.println("Celular apagado.");
        } else {
            System.out.println("Usando " + nombre + " por " + minutos + " minutos. Batería restante: " + String.format("%.2f", bateria) + "%");
        }
    }

    public void mostrarBateria() {
        System.out.println("Batería restante: " + String.format("%.2f", bateria) + "%");
    }
}

File: INTRODUCCION-POO/EJERCICIO-7/PYTHON/Celular.py
class Celular:
    def __init__(self):
        self.aplicaciones = []
        self.espacio_usado = 0
        self.espacio_total = 1024
        self.bateria = 100

    def instalar_aplicacion(self, nombre, tamaño):
        if len(self.aplicaciones) >= 20:
            print("No se pueden instalar más de 20 aplicaciones.")
            return
        if self.espacio_usado + tamaño > self.espacio_total:
            print("No hay suficiente espacio para instalar la aplicación.")
            return

        self.aplicaciones.append(nombre)
        self.espacio_usado += tamaño
        print(f"Aplicación {nombre} instalada con éxito.")

    def usar_aplicacion(self, nombre, minutos):
        if self.bateria == 0:
            print("Celular apagado. Por favor, cargue la batería.")
            return

        for app in self.aplicaciones:
            if app['nombre'] == nombre:
                tamaño = app['tamaño']
                break
        else:
            print("La aplicación no está instalada.")
            return

        if tamaño > 250:
            consumo_bateria = 0.5 * (minutos // 10)
        elif tamaño > 100:
            consumo_bateria = 0.02 * (minutos // 10)
        else:
            consumo_bateria = 0.01 * (minutos // 10)

        self.bateria -= consumo_bateria * 100
        if self.bateria < 0:
            self.bateria = 0

        if self.bateria == 0:
            print("Celular apagado.")
        else:
            print(f"Usando {nombre} por {minutos} minutos. Batería restante: {self.bateria:.2f}%")

    def mostrar_bateria(self):
        print(f"Batería restante: {self.bateria:.2f}%")

celular = Celular()
celular.instalar_aplicacion({"nombre": "WhatsApp", "tamaño": 150})
celular.instalar_aplicacion({"nombre": "Instagram", "tamaño": 300})
celular.usar_aplicacion("WhatsApp", 30)
celular.mostrar_bateria()
celular.usar_aplicacion("Instagram", 20)
celular.mostrar_bateria()
